src/Core/Configurations/ConfigFileWatcher.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.IO.Abstractions;
+using Azure.DataApiBuilder.Config;
+using Path = System.IO.Path;
+
+namespace Azure.DataApiBuilder.Core.Configurations;
+
+public class ConfigFileWatcher
+{
+    private FileSystemWatcher? _fileWatcher;
+    private readonly RuntimeConfigProvider? _configProvider;
+
+    public ConfigFileWatcher(RuntimeConfigProvider configProvider)
+    {
+        FileSystemRuntimeConfigLoader loader = (FileSystemRuntimeConfigLoader)configProvider.ConfigLoader;
+        string configFilePath = loader.ConfigFilePath;
+        IFileSystem fileSystem = (IFileSystem)loader.FileSystem;
+        string? directoryName = Path.GetDirectoryName(configFilePath);
+        directoryName = string.IsNullOrWhiteSpace(directoryName) ?
+            fileSystem.Directory.GetCurrentDirectory() :
+            directoryName;
+
+        _fileWatcher = new FileSystemWatcher(directoryName)
+        {
+            Filter = Path.GetFileName(configFilePath),
+            EnableRaisingEvents = true
+        };
+
+        _configProvider = configProvider;
+        _fileWatcher.Changed += OnConfigFileChange;
+    }
+
+    public ConfigFileWatcher() { }
+
+    /// <summary>
+    /// When a change is detected in the Config file being watched this trigger
+    /// function is called and handles the hot reload logic when appropriate,
+    /// ie: in a local development scenario.
+    /// </summary>
+    /// <param name="sender"></param>
+    /// <param name="e"></param>
+    private void OnConfigFileChange(object sender, FileSystemEventArgs e)
+    {
+        try
+        {
+            if (_configProvider is null)
+            {
+                throw new ArgumentNullException("_configProvider can not be null.");
+            }
+
+            if (!_configProvider!.IsLateConfigured && _configProvider!.GetConfig().IsDevelopmentMode())
+            {
+                _configProvider!.HotReloadConfig();
+            }
+        }
+        catch (Exception ex)
+        {
+            // Need to remove the dependency configuring authentication has on the RuntimeConfigProvider
+            // before we can have an ILogger here.
+            Console.WriteLine("Unable to Hot Reload configuration file due to " + ex.Message);
```

Comment:
In this case you will need to get the config back into a healthy state. So as long as you are trying to hot reload or otherwise load in an unhealthy configuration, you will run into errors with loading the runtime config and get messages stating as such. However, you can recover from this simply by providing a health configuration file, and every way that I have been able to break the hot reloading by getting an exception to trigger has had pretty clear error messaging.

But it is true that if you hot reload with a bad config file, the service will not work until you provide something to get the states back to be consistent and healthy. This is something that was considered early on, and we decided to just let people hot reload as they desire, and then they can reset the state, rather than trying to see if we can reload and if not return it to the last healthy state for them. So this was a choice that we made intentionally.

Reply:
Just to clarify what you'd see with an example:

Let's say I have a healthy config but then I do a bad hot reload. Suppose I put in some invalid characters so that we throw an exception when we try to parse the config during the hot reload call.

This will fail and I will have error messages that relate to the way that the loading of the config during hot reload went bad.  Then if I send a request, since the runtime config isn't setup correctly, it will try to set it up again, but since we havent fixed the broken config, it will fail and you will get an error in the response to your request.

{
    "error": {
        "code": "ErrorInInitialization",
        "message": "Runtime config isn't setup.",
        "status": 503
    }
}